In [12]:
# Lets assume we have a very simple cost function.
def f(x):
    return x * x - 2 * x + 1

# And its corresponding derivative.
def g(x):
    return 2 * x - 2

# Then we can come out the gradient descent algorithm.
def gd(x, alpha, g):
    for i in range(20):
        grad = g(x)
        x -= alpha * grad
        
        print('[Epoch {}] grad = {}, x = {}'.format(i, grad, x))
        
        if abs(grad) < 1e-6:
            break
    return x

For the above $f(x)$ we clearly know it reaches its minimum value at $x = 1$, but lets use our 'gd' algorithm to figure the $x$ out

In [13]:
gd(5, 0.1, g)

[Epoch 0] grad = 8, x = 4.2
[Epoch 1] grad = 6.4, x = 3.56
[Epoch 2] grad = 5.12, x = 3.048
[Epoch 3] grad = 4.096, x = 2.6384
[Epoch 4] grad = 3.2767999999999997, x = 2.31072
[Epoch 5] grad = 2.6214399999999998, x = 2.0485759999999997
[Epoch 6] grad = 2.0971519999999995, x = 1.8388607999999997
[Epoch 7] grad = 1.6777215999999995, x = 1.6710886399999998
[Epoch 8] grad = 1.3421772799999996, x = 1.536870912
[Epoch 9] grad = 1.0737418239999998, x = 1.4294967295999998
[Epoch 10] grad = 0.8589934591999997, x = 1.34359738368
[Epoch 11] grad = 0.6871947673599998, x = 1.274877906944
[Epoch 12] grad = 0.5497558138879999, x = 1.2199023255552
[Epoch 13] grad = 0.4398046511103999, x = 1.17592186044416
[Epoch 14] grad = 0.35184372088831983, x = 1.1407374883553278
[Epoch 15] grad = 0.2814749767106557, x = 1.1125899906842622
[Epoch 16] grad = 0.22517998136852446, x = 1.0900719925474098
[Epoch 17] grad = 0.18014398509481966, x = 1.0720575940379278
[Epoch 18] grad = 0.14411518807585555, x = 1.057646075

1.0461168601842739

Looks good, the $x$ is gradually approaching 1 from its initial value we set 5.

But it took 20 iterations to reach 1.0461168601842739, how about lets increase the learning rate for making it reach there faster, lets set the learning rate 100!!!

In [14]:
gd(5, 100, g)

[Epoch 0] grad = 8, x = -795
[Epoch 1] grad = -1592, x = 158405
[Epoch 2] grad = 316808, x = -31522395
[Epoch 3] grad = -63044792, x = 6272956805
[Epoch 4] grad = 12545913608, x = -1248318403995
[Epoch 5] grad = -2496636807992, x = 248415362395205
[Epoch 6] grad = 496830724790408, x = -49434657116645595
[Epoch 7] grad = -98869314233291192, x = 9837496766212473605
[Epoch 8] grad = 19674993532424947208, x = -1957661856476282247195
[Epoch 9] grad = -3915323712952564494392, x = 389574709438780167192005
[Epoch 10] grad = 779149418877560334384008, x = -77525367178317253271208795
[Epoch 11] grad = -155050734356634506542417592, x = 15427548068485133400970550405
[Epoch 12] grad = 30855096136970266801941100808, x = -3070082065628541546793139530395
[Epoch 13] grad = -6140164131257083093586279060792, x = 610946331060079767811834766548805
[Epoch 14] grad = 1221892662120159535623669533097608, x = -121578319880955873794555118543211995
[Epoch 15] grad = -243156639761911747589110237086423992, x = 24194

37942113558577498272530161493678745851550384005

Above behaviour is also easy to understand: according to the definition of gradient, if the learning rate/step is too big, then probably we will overpass the minimum point, the value of the function will increase instead of decrease.

With a bit more thinking: since smaller learning rate will make the function converge, and bigger learning rate will make the function diverge, then there should be transition point in between will make the function just move in circle.

Lets still start from 5, and choose a special learning rate to make it after two iterations go back to 5 again, lets construct a system of equations:

$$
\begin{cases}
x = 5 &\Rightarrow g(5) = 2 * 5 - 2 = 8 \\
\text{next x: }x' = x - \alpha * g(5) = 5 - 8\alpha &\Rightarrow g(x') = 2 * (5 - 8\alpha) - 2 \\
\text{next next x(and we make it be 5 again): }x'' = x' - \alpha * g(x') = 16\alpha^2 - 16\alpha + 5 = 5 &\Rightarrow \alpha = 1
\end{cases}
$$

That is to say, when learning rate $\alpha = 1$, the iteration will go in circle, lets try:

In [15]:
gd(5, 1, g)

[Epoch 0] grad = 8, x = -3
[Epoch 1] grad = -8, x = 5
[Epoch 2] grad = 8, x = -3
[Epoch 3] grad = -8, x = 5
[Epoch 4] grad = 8, x = -3
[Epoch 5] grad = -8, x = 5
[Epoch 6] grad = 8, x = -3
[Epoch 7] grad = -8, x = 5
[Epoch 8] grad = 8, x = -3
[Epoch 9] grad = -8, x = 5
[Epoch 10] grad = 8, x = -3
[Epoch 11] grad = -8, x = 5
[Epoch 12] grad = 8, x = -3
[Epoch 13] grad = -8, x = 5
[Epoch 14] grad = 8, x = -3
[Epoch 15] grad = -8, x = 5
[Epoch 16] grad = 8, x = -3
[Epoch 17] grad = -8, x = 5
[Epoch 18] grad = 8, x = -3
[Epoch 19] grad = -8, x = 5


5

Just happended like we expected.

So we can conclude for cost function $f(x) = x^2 - 2x + 1$ the transition point for learning rate is 1, smaller than 1 the function will converge, bigger than 1 the function will deverge.

But what is the learning rate transition point for other cost functions, lets do some other experiments:

In [16]:
# Lets assume we have another cost function.
def f1(x):
    return 4 * x * x - 4 * x + 1

# Then its corresponding derivative.
def g1(x):
    return 8 * x - 4

gd(5, 0.25, g1)

[Epoch 0] grad = 36, x = -4.0
[Epoch 1] grad = -36.0, x = 5.0
[Epoch 2] grad = 36.0, x = -4.0
[Epoch 3] grad = -36.0, x = 5.0
[Epoch 4] grad = 36.0, x = -4.0
[Epoch 5] grad = -36.0, x = 5.0
[Epoch 6] grad = 36.0, x = -4.0
[Epoch 7] grad = -36.0, x = 5.0
[Epoch 8] grad = 36.0, x = -4.0
[Epoch 9] grad = -36.0, x = 5.0
[Epoch 10] grad = 36.0, x = -4.0
[Epoch 11] grad = -36.0, x = 5.0
[Epoch 12] grad = 36.0, x = -4.0
[Epoch 13] grad = -36.0, x = 5.0
[Epoch 14] grad = 36.0, x = -4.0
[Epoch 15] grad = -36.0, x = 5.0
[Epoch 16] grad = 36.0, x = -4.0
[Epoch 17] grad = -36.0, x = 5.0
[Epoch 18] grad = 36.0, x = -4.0
[Epoch 19] grad = -36.0, x = 5.0


5.0

PS: looks there are some connections between the learning rate transition point(safety threshold) and the second derivative of the cost function.

### Final question
Must there be a learning rate transition point(safety threshold) for all kinds of cost functions?

But at least in the future we can try to follow the above calculation way to try to find out the learning rate transition point(safety threshold) when we have one actual cost function.

### Credit to [梯度下降是门手艺活……](https://zhuanlan.zhihu.com/p/21486804)